devito/tools/data_structures.py

```diff
@@ -188,3 +188,119 @@ def relations(self):
 
     def generate_ordering(self):
         raise NotImplementedError
+
+
+class DAG(object):
+    """
+    A trivial implementation of a directed acyclic graph (DAG).
+
+    Originally extracted from:
+
+        https://github.com/thieman/py-dag/
```

Comment:
Is this tested?

Reply:
well it's tested [here](https://github.com/thieman/py-dag/blob/master/tests/init_tests.py)

Our implementation is basically a subset of the linked one so...